File: 2024/March/3-5/MinStringDeleteEnds.js
/**
 * @param {string} s
 * @return {number}
 */
var minimumLength = function(s) {
    var pre = 0;
    var suf = s.length - 1;

    if(s[pre] != s[suf]) {
        return s.length;
    }

    curChar = s[pre];
    while(pre < suf) {
        while(pre < suf && s[pre] == curChar) {
            pre++;
        }

        if(pre == suf) {
            return 0;
        }

        while(pre < suf && s[suf] == curChar) {
            suf--;
        }

        if(s[pre] == s[suf]) {
            curChar = s[pre];
        }
        else {
            return suf - pre + 1;
        }
    }

    return suf - pre + 1;
};

File: 2024/March/3-5/MinStringDeleteEnds.py
class Solution:
    def minimumLength(self, s: str) -> int:
        pre = 0
        suf = len(s) - 1

        if s[pre] != s[suf]:
            return len(s)

        curChar = s[pre]
        while pre < suf:
            while pre < suf and s[pre] == curChar:
                pre += 1
            
            if pre == suf:
                return 0
            
            while pre < suf and s[suf] == curChar:
                suf -= 1

            if s[pre] == s[suf]:
                curChar = s[pre]
            else:
                return suf - pre + 1

        return suf - pre + 1
        